src/vsphere_cpi/spec/integration/drs_rules_spec.rb

```diff
@@ -293,7 +293,9 @@
                 {'key' => 'value'}
               )
               vms.push(failed_vm)
-            end.to raise_error(/.*DRS cannot find a host to power on or migrate the virtual machine.*/)
+            end.to raise_error { |error|
+              expect(error.fault).to be_kind_of(VimSdk::Vim::Fault::RuleViolation).or be_kind_of(VimSdk::Vim::Fault::GenericDrsFault)
```

Comment:
Since you're going to the trouble of defining your own `Exception` class then this logic probably belongs in it. Assuming your `VMPowerOnError` has a method called `unplaceable?` then you can match this with:

```ruby
expect do
...
end.to raise_error(VMPowerOnError, be_unplaceable)
```

Reply:
raise_error expect a string as second argument. Are you sure it is possible to write a syntax like this. I tried and it does not work.

Reply:
If it doesn't work then just use the block form as before. Either way the logic should belong in `VMPoserOnError` itself. But according to the RSpec documentation it should work as described here:

https://relishapp.com/rspec/rspec-expectations/docs/composing-matchers

With the predicate matchers as described here:

https://relishapp.com/rspec/rspec-expectations/v/3-7/docs/built-in-matchers/predicate-matchers

Reply:
When I try predicate and compose matcher as you described it gives me error like 
```
 expected VSphereCloud::VMPowerOnError with be unplaceable, got #<VSphereCloud::VMPowerOnError: Could not power on VM '<[Vim.VirtualMachine] vm-109>': This operation would violate a virtual machine affinity/anti-affinity rule.> with backtrace:
```
It is taking be_unplaceable as a string. 
I read the links you posted and it should work according to them. The raise_error composing matcher example is like below 
```
RSpec.describe "Passing matchers to raise_error" do
  specify "you can pass a matcher in place of the message" do
    expect {
      raise RuntimeError, "this goes boom"
    }.to raise_error(RuntimeError, a_string_ending_with("boom"))
  end
end
```
Are we understanding something wrong here?